### PR TITLE
Smooth Speed Calculation

### PR DESCRIPTION
**Pull request by thehanggit**
Generate smooth speed calculation including g factor, p factor, preliminary speed and smooth speed. I followed the instructions steps by steps from issue #386. 

Note that the smooth speed has not been referenced or adopt by any other queries. Will update the weighted speed after QA/QC.

### CODE REVIEW

**Comment by JamesSLogan** [comment]
Thanks for working this out @thehanggit! I tried running the smoothing model against all the data in production and got an error indicating that snowflake ran out of memory. The CI job also hit this same error. I think the recursive CTE approach may not work for us after all. Could you email me some times you're available so we can discuss alternative strategies?

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            22)
+            as g_factor
+    from free_speed
+),
+
+/* Calculate exponential filter denoted as p factor */
+p_factor_value as (
+    select
+        *,
+        volume_sum / (volume_sum + 50) as p_factor
```
Comment (JamesSLogan) [inline transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql:219]: Should 50 be a variable?

**Review by JamesSLogan — COMMENTED** [review]
Thanks again Hang - the model looks great! I think there are some opportunities to add numbers as variables to `dbt_profile.yml` but will leave some of them up to you. Not all numbers need to be added - I think 22 is a good example where a text-based name will help future readers understand what 22 actually represents.

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+hourly_g_factor as (
+    select
+        *,
+        coalesce(
+            avg(case
+                when occupancy_avg < occupancy_threshold
+                    then occupancy_avg / nullifzero(volume_sum) * free_flow_speed * 440
+            end)
+                over (partition by detector_id, week_start, hour),
+            22)
```
[the remark this answers is not in the document]
Comment (thehanggit) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql:210]: Yes, James. 22 is a constant indicating a vehicle effective length of 22 feet. I will make it as a variable and will change all 22 accordingly.

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+),
+
+/* Calculate hourly based g factor, set as 22 if there is a null dataset in the mean function */
+hourly_g_factor as (
+    select
+        *,
+        coalesce(
+            avg(case
+                when occupancy_avg < occupancy_threshold
+                    then occupancy_avg / nullifzero(volume_sum) * free_flow_speed * 440
```
[the remark this answers is not in the document]
Comment (thehanggit) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql:207]: I think so, it is a conversion to get miles per hour (5280 ft / mile and 12 -- 5-minute intervals in an hour). I will make it as a variable and change all 440 accordingly.

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+--     from
+--         speed_preliminary_value as spv
+--     cross join lateral (
+--         select
+--             public.exponential_smooth(spv.speed_preliminary, spv.p_factor::float)
+--                 over (partition by spv.detector_id order by spv.sample_timestamp) as exponential_smooth
+--         from
+--             dual
+--     ) as lateral_result on true
+-- )
```
[the remark this answers is not in the document]
Comment (thehanggit) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql:244]: Make them removed

File: transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql (hunk 1 above)
Comment (thehanggit) [reply transform/models/intermediate/clearinghouse/int_clearinghouse__detector_g_factor_based_speed.sql]: Yes, I made it as a variable and put it in dbt_project yml.

File: transform/dbt_project.yml [hunk 5]
```diff
@@ -53,6 +56,8 @@ vars:
   noon_shift_end: "'14:59:59'"
   pm_shift_start: "'15:00:00'"
   pm_shift_end: "'20:00:00'"
+  vehicle effective length: 22
```
Comment (JamesSLogan) [inline transform/dbt_project.yml:59]: Please replace spaces with underscores 👍

**Review by JamesSLogan — APPROVED** [review]
One last tweak then everything is good to go from my perspective 🚀

**Pull request merged**